### PR TITLE
[4] fix missing com_modules.module.90 from #__assets table

### PR DESCRIPTION
**Pull request by alikon**
### Summary of Changes
added the missed node to the `#__assets` table

### Testing Instructions
code review
or 
enable the log 

### Actual result BEFORE applying this Pull Request
image

### Expected result AFTER applying this Pull Request
asset found for module Login Support

### CODE REVIEW

File: installation/sql/mysql/base.sql [hunk 1]
```diff
@@ -105,7 +105,8 @@ INSERT INTO `#__assets` (`id`, `parent_id`, `lft`, `rgt`, `level`, `name`, `titl
 (85, 18, 120, 121, 2, 'com_modules.module.108', 'Privacy Status', '{}'),
 (86, 18, 122, 123, 2, 'com_modules.module.96', 'Popular Articles', '{}'),
 (87, 18, 124, 125, 2, 'com_modules.module.97', 'Recently Added Articles', '{}'),
-(88, 18, 126, 127, 2, 'com_modules.module.98', 'Logged-in Users', '{}');
+(88, 18, 126, 127, 2, 'com_modules.module.98', 'Logged-in Users', '{}'),
+(89, 18, 129, 130, 2, 'com_modules.module.90', 'Login Support', '{}');
```
Comment (richard67) [inline installation/sql/mysql/base.sql]:
```suggestion
(89, 18, 128, 129, 2, 'com_modules.module.90', 'Login Support', '{}');
```

File: installation/sql/mysql/base.sql [hunk 2]
```diff
@@ -42,23 +42,23 @@ INSERT INTO `#__assets` (`id`, `parent_id`, `lft`, `rgt`, `level`, `name`, `titl
...
-(21, 1, 135, 136, 1, 'com_redirect', 'com_redirect', '{"core.admin":{"7":1}}'),
-(23, 1, 137, 138, 1, 'com_templates', 'com_templates', '{"core.admin":{"7":1}}'),
-(24, 1, 143, 146, 1, 'com_users', 'com_users', '{"core.admin":{"7":1}}'),
-(26, 1, 147, 148, 1, 'com_wrapper', 'com_wrapper', '{}'),
+(19, 1, 131, 134, 1, 'com_newsfeeds', 'com_newsfeeds', '{"core.admin":{"7":1},"core.manage":{"6":1}}'),
+(20, 1, 135, 136, 1, 'com_plugins', 'com_plugins', '{"core.admin":{"7":1}}'),
+(21, 1, 137, 138, 1, 'com_redirect', 'com_redirect', '{"core.admin":{"7":1}}'),
+(23, 1, 139, 140, 1, 'com_templates', 'com_templates', '{"core.admin":{"7":1}}'),
+(24, 1, 145, 148, 1, 'com_users', 'com_users', '{"core.admin":{"7":1}}'),
+(26, 1, 149, 150, 1, 'com_wrapper', 'com_wrapper', '{}'),
```
Comment (richard67) [inline installation/sql/mysql/base.sql:50]:
```suggestion
(19, 1, 130, 133, 1, 'com_newsfeeds', 'com_newsfeeds', '{"core.admin":{"7":1},"core.manage":{"6":1}}'),
(20, 1, 134, 135, 1, 'com_plugins', 'com_plugins', '{"core.admin":{"7":1}}'),
(21, 1, 136, 137, 1, 'com_redirect', 'com_redirect', '{"core.admin":{"7":1}}'),
(23, 1, 138, 139, 1, 'com_templates', 'com_templates', '{"core.admin":{"7":1}}'),
(24, 1, 144, 147, 1, 'com_users', 'com_users', '{"core.admin":{"7":1}}'),
(26, 1, 148, 149, 1, 'com_wrapper', 'com_wrapper', '{}'),
```

File: installation/sql/mysql/base.sql [hunk 3]
```diff
@@ -42,23 +42,23 @@ INSERT INTO `#__assets` (`id`, `parent_id`, `lft`, `rgt`, `level`, `name`, `titl
...
-(37, 1, 157, 158, 1, 'com_ajax', 'com_ajax', '{}'),
-(38, 1, 159, 160, 1, 'com_postinstall', 'com_postinstall', '{}'),
+(30, 19, 132, 133, 2, 'com_newsfeeds.category.5', 'Uncategorised', '{}'),
+(32, 24, 146, 147, 2, 'com_users.category.7', 'Uncategorised', '{}'),
+(33, 1, 151, 152, 1, 'com_finder', 'com_finder', '{"core.admin":{"7":1},"core.manage":{"6":1}}'),
+(34, 1, 153, 154, 1, 'com_joomlaupdate', 'com_joomlaupdate', '{}'),
+(35, 1, 155, 156, 1, 'com_tags', 'com_tags', '{}'),
+(36, 1, 157, 158, 1, 'com_contenthistory', 'com_contenthistory', '{}'),
+(37, 1, 159, 160, 1, 'com_ajax', 'com_ajax', '{}'),
+(38, 1, 161, 162, 1, 'com_postinstall', 'com_postinstall', '{}'),
```
Comment (richard67) [inline installation/sql/mysql/base.sql:61]:
```suggestion
(30, 19, 131, 132, 2, 'com_newsfeeds.category.5', 'Uncategorised', '{}'),
(32, 24, 145, 146, 2, 'com_users.category.7', 'Uncategorised', '{}'),
(33, 1, 150, 151, 1, 'com_finder', 'com_finder', '{"core.admin":{"7":1},"core.manage":{"6":1}}'),
(34, 1, 152, 153, 1, 'com_joomlaupdate', 'com_joomlaupdate', '{}'),
(35, 1, 154, 155, 1, 'com_tags', 'com_tags', '{}'),
(36, 1, 156, 157, 1, 'com_contenthistory', 'com_contenthistory', '{}'),
(37, 1, 158, 159, 1, 'com_ajax', 'com_ajax', '{}'),
(38, 1, 160, 161, 1, 'com_postinstall', 'com_postinstall', '{}'),
```

File: installation/sql/mysql/base.sql [hunk 4]
```diff
@@ -84,8 +84,8 @@ INSERT INTO `#__assets` (`id`, `parent_id`, `lft`, `rgt`, `level`, `name`, `titl
 (62, 56, 31, 32, 3, 'com_content.transition.5', 'Feature', '{}'),
 (63, 56, 33, 34, 3, 'com_content.transition.6', 'Unfeature', '{}'),
 (64, 56, 35, 36, 3, 'com_content.transition.7', 'Publish & Feature', '{}'),
-(65, 1, 139, 140, 1, 'com_privacy', 'com_privacy', '{}'),
-(66, 1, 141, 142, 1, 'com_actionlogs', 'com_actionlogs', '{}'),
+(65, 1, 141, 142, 1, 'com_privacy', 'com_privacy', '{}'),
+(66, 1, 143, 144, 1, 'com_actionlogs', 'com_actionlogs', '{}'),
```
Comment (richard67) [inline installation/sql/mysql/base.sql:88]:
```suggestion
(65, 1, 140, 141, 1, 'com_privacy', 'com_privacy', '{}'),
(66, 1, 142, 143, 1, 'com_actionlogs', 'com_actionlogs', '{}'),
```

**Comment by richard67** [comment]
@alikon I've made suggestions for the mysql file regarding lft and rgt. The same should also be changed to the postgresql file.

**Comment by alikon** [comment]
@richard67 
as we are adding a child to a node that already have childs

i've followed these steps:

1) identify the rgt value of the parent
```sql
SELECT @myRight := rgt FROM `#__assets` where name ='com_modules' 
```

result `@myRight = 128`

2) make room for the new child
```sql
UPDATE `#__assets` SET rgt = rgt + 2 WHERE rgt > @myRight;
UPDATE `#__assets` SET lft = lft + 2 WHERE lft > @myRight;
```

3) add the new child to the parent
```sql
INSERT INTO `#__assets` (`id`, `parent_id`, `lft`, `rgt`, `level`, `name`, `title`, `rules`)
VALUES(89, 18, @myRight + 1, @myRight + 2, 2, 'com_modules.module.90', 'Login Support', '{}');
```

**Comment by richard67** [comment]
@alikon The procedure seems right to me. Maybe I've made a mistake when checking manually. Will check again later.

File: installation/sql/mysql/base.sql (hunk 1 above)
Comment (alikon) [reply installation/sql/mysql/base.sql]: we cannot re-use 128 as 128 is the the rgt value of the parent

Comment (richard67) [reply installation/sql/mysql/base.sql]: Correct, I've missed the 128.

**Comment by richard67** [comment]
Hmm, but shouldn't the rgt value of record 18 (com_modules) be changed, too, when adding new items with parent = 18?

File: installation/sql/mysql/base.sql (hunk 1 above)
Comment (richard67) [reply installation/sql/mysql/base.sql]: No, now you confused me. The point is that you have to change the rgt=128 of the record 18, too, which is the parent.

The last child of record 18 on level 2 is record 88, which has lft=126 and rgt=127.

Now we append the new item to that on the same level 2. So the new item gets lft=128 and rgt = 129, like I suggested, and the other childs get 2 more for lft and rgt than they had before, and then the rgt of the parent gets also 2 more, and that is the record 18, so it gets rgt=130 now.

**Comment by richard67** [comment]
I think step 2 should be:

```
UPDATE `#__assets` SET rgt = rgt + 2 WHERE rgt >= @myRight;
UPDATE `#__assets` SET lft = lft + 2 WHERE lft > @myRight;
```

I.e. for the rgt value use ">=" and for the lft value ">" in the condition.

Then the parent's rgt value will be updated in the right way, too.

Or you leave these conditions like you wrote but have to update that one record later.

Update: But the ">=" is only true if you insert a child of that record for which you got `@myRight`. If you insert on the same level 1 and not like here level 2, it would be the ">".

**Comment by richard67** [comment]
Ok, I think I have it. Will mark my previous suggestions as resolved and make a new one.

File: installation/sql/mysql/base.sql (hunk 1 above)
Comment (richard67) [reply installation/sql/mysql/base.sql]: Will close this suggestion and the others and make a new one.

File: installation/sql/postgresql/base.sql [hunk 5]
```diff
@@ -111,9 +111,10 @@ INSERT INTO "#__assets" ("id", "parent_id", "lft", "rgt", "level", "name", "titl
 (85, 18, 120, 121, 2, 'com_modules.module.108', 'Privacy Status', '{}'),
 (86, 18, 122, 123, 2, 'com_modules.module.96', 'Popular Articles', '{}'),
 (87, 18, 124, 125, 2, 'com_modules.module.97', 'Recently Added Articles', '{}'),
-(88, 18, 126, 127, 2, 'com_modules.module.98', 'Logged-in Users', '{}');
+(88, 18, 126, 127, 2, 'com_modules.module.98', 'Logged-in Users', '{}'),
+(89, 18, 129, 130, 2, 'com_modules.module.90', 'Login Support', '{}');
```
Comment (richard67) [inline installation/sql/postgresql/base.sql]:
```suggestion
(89, 18, 128, 129, 2, 'com_modules.module.90', 'Login Support', '{}');
```

File: installation/sql/postgresql/base.sql [hunk 6]
```diff
@@ -48,23 +48,23 @@ INSERT INTO "#__assets" ("id", "parent_id", "lft", "rgt", "level", "name", "titl
 (16, 1, 51, 54, 1, 'com_menus', 'com_menus', '{"core.admin":{"7":1}}'),
 (17, 1, 55, 56, 1, 'com_messages', 'com_messages', '{"core.admin":{"7":1},"core.manage":{"7":1}}'),
 (18, 1, 57, 128, 1, 'com_modules', 'com_modules', '{"core.admin":{"7":1}}'),
```
Comment (richard67) [inline installation/sql/postgresql/base.sql]:
```suggestion
(18, 1, 57, 130, 1, 'com_modules', 'com_modules', '{"core.admin":{"7":1}}'),
```

File: installation/sql/mysql/base.sql (hunk 1 above)
Comment (richard67) [inline installation/sql/mysql/base.sql]:
```suggestion
(89, 18, 128, 129, 2, 'com_modules.module.90', 'Login Support', '{}');
```

File: installation/sql/mysql/base.sql [hunk 7]
```diff
@@ -42,23 +42,23 @@ INSERT INTO `#__assets` (`id`, `parent_id`, `lft`, `rgt`, `level`, `name`, `titl
 (16, 1, 51, 54, 1, 'com_menus', 'com_menus', '{"core.admin":{"7":1}}'),
 (17, 1, 55, 56, 1, 'com_messages', 'com_messages', '{"core.admin":{"7":1},"core.manage":{"7":1}}'),
 (18, 1, 57, 128, 1, 'com_modules', 'com_modules', '{"core.admin":{"7":1}}'),
```
Comment (richard67) [inline installation/sql/mysql/base.sql]:
```suggestion
(18, 1, 57, 130, 1, 'com_modules', 'com_modules', '{"core.admin":{"7":1}}'),
```

**Review by richard67 — CHANGES_REQUESTED** [review]
Now my suggestion should be right. It was all ok wqith your PR except of the lft and rgt of the inserted item and the rgt of the parent (record 18).

**Comment by alikon** [comment]
ok so let's review it  for this pr and for future reference

as we are adding a child to a parent that already have childs

these steps needs to be done:

1) identify the `rgt` value of the parent and take note of the `parent_id` as well
```sql
SELECT @myRight := rgt FROM `#__assets` where name ='com_modules' 
```

result `@myRight = 128`

2) make room for the new child
```sql
UPDATE `#__assets` SET rgt = rgt + 2 WHERE rgt >= @myRight;
UPDATE `#__assets` SET lft = lft + 2 WHERE lft > @myRight;
```

3) add the new child to the parent
```sql
INSERT INTO `#__assets` (`id`, `parent_id`, `lft`, `rgt`, `level`, `name`, `title`, `rules`)
VALUES(89, 18, @myRight , @myRight + 1, 2, 'com_modules.module.90', 'Login Support', '{}');
```

4) check that the new child is a child of the parent

```sql
SELECT node.name, node.parent_id
FROM `#__assets` AS node,
        `#__assets` AS parent
WHERE node.lft BETWEEN parent.lft AND parent.rgt
        AND parent.name = 'com_modules'
ORDER BY node.lft
```

**Comment by richard67** [comment]
@alikon Yes that looks ok to me. The result should be the same as my latest review suggestion.

**Comment by richard67** [comment]
@alikon Normally (in a world where the features of relational databases are really used), an insert trigger would do that for us.

Unfortunately as we support 2 kinds of databases, it would mean we either maintain triggers for each database kind in their programming language, or we implement the triggers in external libraries in the programming language of our choice and let the database triggers use these external procedures. Both would be possible.

**Review by richard67 — APPROVED** [review]

**Review by richard67 — APPROVED** [review]

**Comment by alikon** [comment]
i would prefer to go with a stored procedure, yeah maybe for J7 or J9 😄

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on d2a7bb11d5805eab7d4b0ccbf53fe7b56288c256<br /><br />Code review + real test.

Hint for other testers for real test: Just check the "Log" tab of the debug when having "Debug System" switched on on any 4.0 or 4.1 installation. You will see the 2 logs about the missing asset for module 90. Then make a new installation with the PR applied and check the same. The log is empty now.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36481">issues.joomla.org/tracker/joomla-cms/36481</a>.</sub>

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**